File: agent.go
package main

import (
	"context"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"markestedt/tokentalk/audio"
	"markestedt/tokentalk/config"
	"markestedt/tokentalk/platform"
	"markestedt/tokentalk/storage"
	"markestedt/tokentalk/transcribe"
	"markestedt/tokentalk/web"
)

// Agent coordinates hotkey detection, recording, and transcription
type Agent struct {
	cfg       *config.Config
	hotkey    platform.Hotkey
	clipboard platform.Clipboard
	paster    platform.Paster
	recorder  *audio.Recorder
	provider  transcribe.Provider
	db        *storage.DB
	webServer *web.Server
}

// NewAgent creates a new agent instance
func NewAgent(cfg *config.Config) (*Agent, error) {
	// Convert device ID to string
	deviceID := ""
	if cfg.Audio.Device > 0 {
		deviceID = fmt.Sprintf("%d", cfg.Audio.Device)
	}

	// Create recorder
	recorder, err := audio.NewRecorder(deviceID, cfg.Audio.MaxSeconds)
	if err != nil {
		return nil, fmt.Errorf("failed to create recorder: %w", err)
	}

	// Create transcription provider
	provider, err := transcribe.NewProvider(cfg.Transcription, cfg.DeveloperMode)
	if err != nil {
		return nil, fmt.Errorf("failed to create transcription provider: %w", err)
	}

	return &Agent{
		cfg:       cfg,
		hotkey:    platform.NewHotkey(),
		clipboard: platform.NewClipboard(),
		paster:    platform.NewPaster(),
		recorder:  recorder,
		provider:  provider,
	}, nil
}

// setStatus updates the agent status and broadcasts to web clients
func (a *Agent) setStatus(status string) {
	if a.webServer != nil {
		a.webServer.BroadcastStatus(status)
	}
}

// Run starts the agent's main event loop
func (a *Agent) Run(ctx context.Context) error {
	// Parse hotkey combo
	combo, err := config.ParseHotkey(a.cfg.Hotkey)
	if err != nil {
		return fmt.Errorf("failed to parse hotkey: %w", err)
	}

	// Convert key to VK code (0 means modifier-only combo)
	vkCode, err := platform.VKCode(combo.Key)
	if err != nil {
		return fmt.Errorf("failed to get VK code: %w", err)
	}

	pkCombo := platform.KeyCombo{
		Ctrl:  combo.Ctrl,
		Shift: combo.Shift,
		Alt:   combo.Alt,
		Win:   combo.Win,
		Key:   vkCode,
	}

	// Start listening for hotkey
	events, err := a.hotkey.Listen(ctx, pkCombo)
	if err != nil {
		return fmt.Errorf("failed to start hotkey listener: %w", err)
	}

	slog.Info("TokenTalk started", "hotkey", a.cfg.Hotkey, "provider", a.provider.Name())
	a.setStatus("idle")

	// Main event loop
	for {
		select {
		case <-ctx.Done():
			a.recorder.Close()
			return nil

		case evt := <-events:
			switch evt.Type {
			case platform.Pressed:
				// Start recording immediately to minimize latency
				if err := a.recorder.Start(ctx); err != nil {
					slog.Error("Failed to start recording", "error", err)
					a.setStatus("idle")
					continue
				}
				// Log and update status asynchronously to avoid blocking
				go func() {
					slog.Info("Recording started")
					a.setStatus("recording")
				}()

			case platform.Released:
				slog.Info("Recording stopped, transcribing...")
				recordingStart := time.Now()
				audioSeg, err := a.recorder.Stop()
				if err != nil {
					slog.Error("Failed to stop recording", "error", err)
					a.setStatus("idle")
					continue
				}

				// Check if audio is too short
				if audioSeg.Duration < 100*time.Millisecond {
					slog.Warn("Recording too short, ignoring", "duration", audioSeg.Duration)
					a.setStatus("idle")
					continue
				}

				// Check if audio is silent or too quiet (if threshold is set)
				if a.cfg.Audio.SilenceThreshold > 0 {
					rms := audioSeg.CalculateRMS()
					if rms < a.cfg.Audio.SilenceThreshold {
						slog.Warn("Recording too quiet or silent, ignoring", "rms", rms, "threshold", a.cfg.Audio.SilenceThreshold)
						a.setStatus("idle")
						continue
					}
				}

				// Transcribe in background to avoid blocking
				go func(seg audio.AudioSegment) {
					a.setStatus("processing")

					dictation := &storage.Dictation{
						RecordingStartMs:    recordingStart.UnixMilli(),
						RecordingDurationMs: seg.Duration.Milliseconds(),
						AudioSizeBytes:      int64(len(seg.Data)),
						AudioSampleRate:     seg.SampleRate,
						Provider:            a.provider.Name(),
						Model:               a.cfg.Transcription.Model,
						Language:            a.cfg.Transcription.Language,
						Success:             false,
					}

					transcribeStart := time.Now()
					text, err := a.provider.Transcribe(ctx, seg)
					dictation.TranscriptionLatencyMs = time.Since(transcribeStart).Milliseconds()

					if err != nil {
						slog.Error("Transcription failed", "error", err)
						dictation.ErrorMessage = err.Error()
						if a.db != nil {
							a.db.SaveDictation(dictation)
						}
						a.setStatus("idle")
						return
					}

					if text == "" {
						slog.Warn("Empty transcription")
						dictation.ErrorMessage = "Empty transcription"
						if a.db != nil {
							a.db.SaveDictation(dictation)
						}
						a.setStatus("idle")
						return
					}

					dictation.TranscribedText = text
					dictation.WordCount = len(strings.Fields(text))
					dictation.CharacterCount = len(text)

					slog.Info("Transcribed", "text", text, "duration", seg.Duration)

					// Inject text
					injectStart := time.Now()
					if err := a.injectText(text); err != nil {
						slog.Error("Failed to inject text", "error", err)
						dictation.ErrorMessage = err.Error()
						dictation.InjectionLatencyMs = time.Since(injectStart).Milliseconds()
						dictation.TotalLatencyMs = time.Since(recordingStart).Milliseconds()
						if a.db != nil {
							a.db.SaveDictation(dictation)
						}
						a.setStatus("idle")
						return
					}

					dictation.InjectionLatencyMs = time.Since(injectStart).Milliseconds()
					dictation.TotalLatencyMs = time.Since(recordingStart).Milliseconds()
					dictation.Success = true

					if a.db != nil {
						if err := a.db.SaveDictation(dictation); err != nil {
							slog.Error("Failed to save dictation", "error", err)
						} else if a.webServer != nil {
							a.webServer.BroadcastDictation(dictation)
						}
					}

					a.setStatus("idle")
				}(audioSeg)
			}
		}
	}
}

// injectText injects transcribed text via clipboard paste
func (a *Agent) injectText(text string) error {
	// Save current clipboard content
	originalClip, err := a.clipboard.Get()
	if err != nil {
		slog.Warn("Failed to get clipboard content, continuing anyway", "error", err)
		originalClip = ""
	}

	// Set clipboard to transcribed text
	if err := a.clipboard.Set(text); err != nil {
		return fmt.Errorf("failed to set clipboard: %w", err)
	}

	// Wait for clipboard to update
	time.Sleep(50 * time.Millisecond)

	// Simulate paste
	if err := a.paster.Paste(); err != nil {
		return fmt.Errorf("failed to paste: %w", err)
	}

	// Wait for paste to complete
	time.Sleep(100 * time.Millisecond)

	// Restore original clipboard
	if originalClip != "" {
		if err := a.clipboard.Set(originalClip); err != nil {
			slog.Warn("Failed to restore clipboard", "error", err)
		}
	}

	return nil
}

File: src/TokenTalk/Agent.cs
using Microsoft.Extensions.Logging;
using TokenTalk.Audio;
using TokenTalk.Configuration;
using TokenTalk.Overlay;
using TokenTalk.Platform;
using TokenTalk.PostProcessing;
using TokenTalk.Storage;
using TokenTalk.Transcription;

namespace TokenTalk;

public class Agent : IDisposable
{
    private readonly ConfigManager _configManager;
    private readonly AudioRecorder _recorder;
    private readonly ITranscriptionProvider _transcriptionProvider;
    private readonly PostProcessingPipeline _pipeline;
    private readonly ClipboardService _clipboard;
    private readonly PasteService _paste;
    private readonly DictationRepository _repository;
    private readonly DictationOverlay? _overlay;
    private readonly HotkeyListener _hotkeyListener;
    private readonly ILogger<Agent> _logger;

    public event EventHandler<string>? StatusChanged;
    public event EventHandler<DictationCompletedEventArgs>? DictationCompleted;

    public Agent(
        ConfigManager configManager,
        AudioRecorder recorder,
        ITranscriptionProvider transcriptionProvider,
        PostProcessingPipeline pipeline,
        ClipboardService clipboard,
        PasteService paste,
        DictationRepository repository,
        DictationOverlay? overlay,
        ILogger<Agent> logger)
    {
        _configManager = configManager;
        _recorder = recorder;
        _transcriptionProvider = transcriptionProvider;
        _pipeline = pipeline;
        _clipboard = clipboard;
        _paste = paste;
        _repository = repository;
        _overlay = overlay;
        _hotkeyListener = new HotkeyListener();
        _logger = logger;

        if (_overlay != null)
            _recorder.AmplitudeAvailable += _overlay.PushAmplitude;
    }

    public async Task RunAsync(CancellationToken ct)
    {
        var cfg = _configManager.Current;

        _hotkeyListener.Start(cfg.Hotkey);
        _logger.LogInformation("TokenTalk started. Hotkey: {Hotkey}, Provider: {Provider}",
            cfg.Hotkey, _transcriptionProvider.Name);

        SetStatus("idle");

        try
        {
            await foreach (var evt in _hotkeyListener.Events.ReadAllAsync(ct))
            {
                switch (evt.Type)
                {
                    case HotkeyEventType.Pressed:
                        HandleHotkeyPressed();
                        break;
                    case HotkeyEventType.Released:
                        _ = HandleHotkeyReleasedAsync(ct);
                        break;
                }
            }
        }
        catch (OperationCanceledException)
        {
            _logger.LogInformation("Agent stopped");
        }
    }

    private void HandleHotkeyPressed()
    {
        try
        {
            _recorder.Start();
            _overlay?.StartRecording();
            _logger.LogInformation("Recording started");
            SetStatus("recording");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to start recording");
            SetStatus("idle");
        }
    }

    private async Task HandleHotkeyReleasedAsync(CancellationToken ct)
    {
        _logger.LogInformation("Recording stopped, transcribing...");
        var recordingStart = DateTimeOffset.UtcNow;

        AudioSegment audio;
        try
        {
            audio = _recorder.Stop();
            _overlay?.StartProcessing();
        }
        catch (Exception ex)
        {
            _overlay?.StopProcessing();
            _logger.LogError(ex, "Failed to stop recording");
            SetStatus("idle");
            return;
        }

        var cfg = _configManager.Current;

        // Validate duration
        if (AudioHelpers.IsTooShort(audio, TimeSpan.FromMilliseconds(100)))
        {
            _logger.LogWarning("Recording too short ({Duration}ms), ignoring", audio.Duration.TotalMilliseconds);
            _overlay?.StopProcessing();
            SetStatus("idle");
            return;
        }

        // Validate silence
        if (cfg.Audio.SilenceThreshold > 0 && AudioHelpers.IsSilent(audio, cfg.Audio.SilenceThreshold))
        {
            _logger.LogWarning("Recording too quiet, ignoring");
            _overlay?.StopProcessing();
            SetStatus("idle");
            return;
        }

        SetStatus("processing");

        var dictation = new Dictation
        {
            RecordingStartMs = recordingStart.ToUnixTimeMilliseconds(),
            RecordingDurationMs = (long)audio.Duration.TotalMilliseconds,
            AudioSizeBytes = audio.WavData.Length,
            AudioSampleRate = audio.SampleRate,
            Provider = _transcriptionProvider.Name,
            Model = cfg.Transcription.Model,
            Language = cfg.Transcription.Language,
            Success = false,
        };

        try
        {
            // Transcribe
            var transcribeStart = DateTimeOffset.UtcNow;
            string text;
            try
            {
                text = await _transcriptionProvider.TranscribeAsync(audio, ct);
                dictation.TranscriptionLatencyMs = (long)(DateTimeOffset.UtcNow - transcribeStart).TotalMilliseconds;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Transcription failed");
                dictation.TranscriptionLatencyMs = (long)(DateTimeOffset.UtcNow - transcribeStart).TotalMilliseconds;
                dictation.ErrorMessage = ex.Message;
                await SaveDictationAsync(dictation, ct);
                SetStatus("idle");
                return;
            }

            if (string.IsNullOrWhiteSpace(text))
            {
                _logger.LogWarning("Empty transcription");
                dictation.ErrorMessage = "Empty transcription";
                await SaveDictationAsync(dictation, ct);
                SetStatus("idle");
                return;
            }

            dictation.TranscribedText = text;
            dictation.WordCount = text.Split(' ', StringSplitOptions.RemoveEmptyEntries).Length;
            dictation.CharacterCount = text.Length;

            _logger.LogInformation("Transcribed: {Text} ({Duration})", text, audio.Duration);

            // Post-process
            var processed = text;
            try
            {
                processed = await _pipeline.ProcessAsync(text, ct);
                if (processed != text)
                    _logger.LogInformation("Post-processed: {Original} → {Processed}", text, processed);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Post-processing failed, using original text");
            }

            // Inject text
            var injectStart = DateTimeOffset.UtcNow;
            try
            {
                await _paste.PasteTextAsync(processed, ct);
                dictation.InjectionLatencyMs = (long)(DateTimeOffset.UtcNow - injectStart).TotalMilliseconds;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to inject text");
                dictation.InjectionLatencyMs = (long)(DateTimeOffset.UtcNow - injectStart).TotalMilliseconds;
                dictation.TotalLatencyMs = (long)(DateTimeOffset.UtcNow - recordingStart).TotalMilliseconds;
                dictation.ErrorMessage = ex.Message;
                await SaveDictationAsync(dictation, ct);
                SetStatus("idle");
                return;
            }

            dictation.TotalLatencyMs = (long)(DateTimeOffset.UtcNow - recordingStart).TotalMilliseconds;
            dictation.Success = true;

            await SaveDictationAsync(dictation, ct);
            DictationCompleted?.Invoke(this, new DictationCompletedEventArgs(dictation));
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Unexpected error in dictation flow");
            dictation.ErrorMessage = ex.Message;
            dictation.TotalLatencyMs = (long)(DateTimeOffset.UtcNow - recordingStart).TotalMilliseconds;
            await SaveDictationAsync(dictation, ct);
        }
        finally
        {
            _overlay?.StopProcessing();
            SetStatus("idle");
        }
    }

    private async Task SaveDictationAsync(Dictation dictation, CancellationToken ct)
    {
        try
        {
            await _repository.SaveAsync(dictation, ct);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to save dictation");
        }
    }

    private void SetStatus(string status)
    {
        StatusChanged?.Invoke(this, status);
    }

    public void Dispose()
    {
        if (_overlay != null)
            _recorder.AmplitudeAvailable -= _overlay.PushAmplitude;
        _hotkeyListener.Dispose();
        _recorder.Dispose();
    }
}

public sealed class DictationCompletedEventArgs : EventArgs
{
    public Dictation Dictation { get; }
    public DictationCompletedEventArgs(Dictation d) => Dictation = d;
}
